File: week09/26/prob2.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> colors(n);

        for (int i = 0; i < n; i++) {
            if (colors[i]) continue;
            queue<int> q{{i}};
            colors[i] = 1;

            while (!q.empty()) {
                int curr = q.front();
                q.pop();
                for (int neighbor : graph[curr]) {
                    if (colors[neighbor] == colors[curr]) return false;
                    if (colors[neighbor] == 0) {
                        colors[neighbor] = -colors[curr];
                        q.push(neighbor);
                    }
                }
            }
        }
        return true;
    }
};

File: week09/26/prob2.py
class Solution(object):
    def isBipartite(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: bool
        """
        if len(graph) == 1:
            return True


        color = [None for _ in graph]
        for i in range(len(graph)):
            if color[i]:
                continue 
            to_visit = [i]
            color[i] = 1
            visited = []
            while to_visit:
                curr = to_visit.pop()
                visited.append(curr)
                for neighbor in graph[curr]:
                    if neighbor not in visited:
                        to_visit.append(neighbor)
                        if color[neighbor] is None:
                            color[neighbor] = -color[curr]
                        else:
                            if color[neighbor] == color[curr]:
                                return False
        return True